Bug 582847 - Add support for Jakarta ejb annotations

loetscher opened the pull request:
Fix for bug https://bugs.eclipse.org/bugs/show_bug.cgi?id=582847

erik-brangs commented on plugins/org.eclipse.jst.j2ee/archiveops/org/eclipse/jst/j2ee/internal/archive/JavaEEArchiveUtilities.java:
```diff
@@ -918,6 +918,14 @@ private static void wrapArchive(final IArchive simpleArchive, final IPath deploy
+
+	private static final char[] MESSAGEDRIVEN_JAKARTA = "Ljakarta/ejb/MessageDriven;".toCharArray();//$NON-NLS-1$
+
+	private static final char[] SINGLETON_JAKARTA = "Ljakarta/ejb/Singleton;".toCharArray();//$NON-NLS-1$
```
I think it would be better to rename the old constants to JAVA_EE_* (e.g. `SINGLETON_JAVA_EE`) and just use the old names for the new constants, i.e. `STATELESS` instead of `STATELESS_JAKARTA`.

loetscher replied:
Renaming is done.

erik-brangs commented on plugins/org.eclipse.jst.j2ee/archiveops/org/eclipse/jst/j2ee/internal/archive/JavaEEArchiveUtilities.java:
```diff
@@ -948,7 +956,8 @@ public boolean isEJBArchive(IArchive archive) {
 											char[] typedName = annotation.getTypeName();
-											if (Arrays.equals(typedName, STATELESS) || Arrays.equals(typedName, STATEFUL) || Arrays.equals(typedName, MESSAGEDRIVEN) || Arrays.equals(typedName, SINGLETON)) {
+											if (Arrays.equals(typedName, STATELESS) || Arrays.equals(typedName, STATEFUL) || Arrays.equals(typedName, MESSAGEDRIVEN) || Arrays.equals(typedName, SINGLETON)
+											    || hasJakartaEJBAnnotation(typedName)) {
```
You could introduce a similar method for Java EE to improve readability. However, I'm not sure which name would be suitable. Here are some proposals: `hasJavaEEEJBAnnotation`, `hasOldStyleEJBAnnotation`, `hasLegacyEJBAnnotation`

loetscher replied:
helper Method introduced

nitind replied:
Generally I've been following a pattern of inserting Jakarta into the name of the parallel new helper methods and constants, e.g. hasEJBAnnotation gets a hasJakartaEJBAnnotation created for that case.

loetscher replied:
There are now two helper methods:
* hasJavaEEEJBAnnotation (legacy annotations for history)
* hasJakartaEJBAnnotation (current Java EE Annotations)
Or statement using this two methods may be switched in future releases for performance optimization (when generally more jakarta.ejb than javax.ejb annotations are used...)

erik-brangs reviewed (commented):
A test case would be nice to have. Are you willing to add one?

erik-brangs commented:
The ECA check is failing. Have you signed the [ECA](https://accounts.eclipse.org/user/eca) with the same email address that you're using for GitHub?

loetscher commented:
> The ECA check is failing. Have you signed the [ECA](https://accounts.eclipse.org/user/eca) with the same email address that you're using for GitHub?

Should be signed now.

erik-brangs reviewed (changes requested):
I think you need to adjust the version in the pom.xml or the build will fail.

The commit messages could also be changed to refer to the bug number (e.g. "Bug 582847 - Add support for Jakarta ejb annotations" or "[582847] - Add support for Jakarta ejb annotations".

erik-brangs commented:
@nitind Could you take a look at this PR, please?

loetscher commented:
> I think you need to adjust the version in the pom.xml or the build will fail.
What should the next version be? A patch increment?

> The commit messages could also be changed to refer to the bug number (e.g. "Bug 582847 - Add support for Jakarta ejb annotations" or "[582847] - Add support for Jakarta ejb annotations".
Rebased the commits and changed commit comment

erik-brangs commented:
> What should the next version be? A patch increment?

I'm using the guidelines from https://wiki.eclipse.org/Version_Numbering . AFAIK those are still uptodate. According to these guidelines, I would say it's a bugfix for the next release so the last segment should be changed. Both the pom.xml and the MANIFEST.mf should be changed to use 1.2.1000 .

loetscher commented:
> > A test case would be nice to have. Are you willing to add one?
> 
> Sure.

* Any hints how to test/mock IArchive Interfaces?
* The testcode should be placed unter `/org.eclipse.jst.j2ee.tests/j2ee-tests/org/eclipse/jst/j2ee/internal/archive/tests/`?

erik-brangs commented:
> The testcode should be placed unter /org.eclipse.jst.j2ee.tests/j2ee-tests/org/eclipse/jst/j2ee/internal/archive/tests/?

Yes, I think that's a good idea. Note that this will be a change to the test plugin so you also need to increment the version like for the original plugin.

> Any hints how to test/mock IArchive Interfaces?

You could try to create an archive via `JavaEEArchiveUtilities.INSTANCE.openArchive`. There are a few examples for this in the tests. There is also `IArchiveFactory.INSTANCE.openArchive` but AFAIK there are no examples for this in the test.

With respect to mocking, you would have to add a dependency for the mocking framework to the plugin. I'm not sure how this works in detail. There is a list of libraries from Orbit at https://download.eclipse.org/tools/orbit/simrel/maven-osgi/2024-03/index.html and it contains a bundle for Mockito with the identifier `org.mockito.mockito-core ` but I don't know if that's available in the build.

loetscher commented:
> You could try to create an archive via `JavaEEArchiveUtilities.INSTANCE.openArchive`. There are a few examples for this in the tests. There is also `IArchiveFactory.INSTANCE.openArchive` but AFAIK there are no examples for this in the test.

I need more time to understand how to test the class `JavaEEArchiveUtilities` in an efficient way.

My suggestion: the testcoverage will stay as it is and I will do the test/refactoring on a separate pr.

erik-brangs commented:
> My suggestion: the testcoverage will stay as it is and I will do the test/refactoring on a separate pr.

That's fine by me.

erik-brangs commented:
@nitind I think this PR is suitable for merging. If you agree, could you merge it, please?

nitind commented:
Approved and merged, thanks!
